verilog/tools/kythe/kythe_facts_extractor.cc

```diff
@@ -0,0 +1,205 @@
+                   module_instance_vname);
+
+  for (const auto& anchor :
+       verible::make_range(anchors.begin() + 2, anchors.end())) {
+    const VName port_vname_reference(
+        file_path_,
+        CreateVariableSignature(anchor.Value(), vnames_context_.top()));
+    const VName port_vname_definition(
+        file_path_,
+        CreateVariableSignature(anchor.Value(), vnames_context_.top()));
+    const VName port_vname_anchor = PrintAnchorVName(anchor, file_path_);
+
+    *stream_ << Edge(port_vname_anchor, kEdgeRef, port_vname_definition);
+  }
+
+  return module_instance_vname;
+}
+
+VName KytheFactsExtractor::ExtractVariableDefinition(
+    const IndexingFactNode& node) {
+  const auto& anchor = node.Value().Anchors()[0];
+  const VName variable_vname(
+      file_path_,
+      CreateVariableSignature(anchor.Value(), vnames_context_.top()));
+  const VName variable_vname_anchor = PrintAnchorVName(anchor, file_path_);
+
+  *stream_ << Fact(variable_vname, kFactNodeKind, kNodeVariable);
+  *stream_ << Fact(variable_vname, kFactComplete, kCompleteDefinition);
+
+  *stream_ << Edge(variable_vname_anchor, kEdgeDefinesBinding, variable_vname);
+
+  return variable_vname;
+}
+
+VName KytheFactsExtractor::ExtractVariableReference(
+    const IndexingFactNode& node) {
+  const auto& anchor = node.Value().Anchors()[0];
+  const VName variable_vname(
+      file_path_,
+      CreateVariableSignature(anchor.Value(), vnames_context_.top()));
+  const VName variable_vname_anchor = PrintAnchorVName(anchor, file_path_);
+
+  *stream_ << Edge(variable_vname_anchor, kEdgeRef, variable_vname);
+
+  return variable_vname;
+}
+
+VName KytheFactsExtractor::PrintAnchorVName(const Anchor& anchor,
+                                            absl::string_view file_path) {
+  const VName anchor_vname(file_path,
+                           absl::Substitute(R"(@$0:$1)", anchor.StartLocation(),
+                                            anchor.EndLocation()));
+
+  *stream_ << Fact(anchor_vname, kFactNodeKind, kNodeAnchor);
+  *stream_ << Fact(anchor_vname, kFactAnchorStart,
+                   absl::Substitute(R"($0)", anchor.StartLocation()));
+  *stream_ << Fact(anchor_vname, kFactAnchorEnd,
+                   absl::Substitute(R"($0)", anchor.EndLocation()));
+
+  return anchor_vname;
+}
+
+std::string CreateModuleSignature(absl::string_view module_name) {
+  return absl::StrCat(module_name, "#module");
+}
+
+std::string CreateVariableSignature(absl::string_view instance_name,
+                                    const VName& parent_vname) {
+  return absl::StrCat(instance_name, "#variable", parent_vname.signature);
+}
+
+std::string GetFilePathFromRoot(const IndexingFactNode& root) {
+  return root.Value().Anchors()[0].Value();
+}
+
+std::ostream& KytheFactsPrinter::Print(std::ostream& stream) const {
+  KytheFactsExtractor kythe_extractor(GetFilePathFromRoot(root_), &stream);
+  kythe_extractor.Visit(root_);
```

Comment:
Canonically, with the visitor pattern, the top-level call is dispatched like:

```c++
root_.Accept(visitor_object);
```

Reply:
root_ is a `VectorTree` and it doesn't have a Accept function.

Reply:
I should not review at 2am.